File: backend/app/services/seo_service.py
"""SEO Service for CyberPress Platform

This module handles SEO-related operations including:
- Meta tag generation
- Sitemap generation
- Robots.txt generation
- Open Graph tags
- Twitter Card tags
- Schema.org structured data
"""

from typing import Dict, Any, List, Optional
from datetime import datetime
import logging

logger = logging.getLogger(__name__)


class SEOService:
    """Service for SEO-related operations"""

    def __init__(self):
        """Initialize SEO service"""
        self.site_name = "CyberPress"
        self.site_url = "https://cyberpress.dev"
        self.default_description = "A cyberpunk-powered blogging platform"

    def generate_meta_tags(
        self,
        title: str,
        description: str,
        keywords: Optional[List[str]] = None,
        image: Optional[str] = None,
        url: Optional[str] = None,
        noindex: bool = False,
        canonical: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Generate meta tags for a page

        Args:
            title: Page title
            description: Page description
            keywords: List of keywords
            image: Featured image URL
            url: Page URL
            noindex: Whether to noindex this page
            canonical: Canonical URL

        Returns:
            Dictionary containing meta tags
        """
        full_title = f"{title} | {self.site_name}" if title != self.site_name else title

        meta_tags = {
            "title": full_title,
            "description": description[:160] if len(description) > 160 else description,
            "keywords": ", ".join(keywords) if keywords else "",
            "canonical": canonical or url,
            "noindex": noindex,
            "open_graph": self._generate_open_graph_tags(
                title=title,
                description=description,
                image=image,
                url=url,
            ),
            "twitter_card": self._generate_twitter_card_tags(
                title=title,
                description=description,
                image=image,
            ),
        }

        return meta_tags

    def _generate_open_graph_tags(
        self,
        title: str,
        description: str,
        image: Optional[str] = None,
        url: Optional[str] = None,
    ) -> Dict[str, str]:
        """Generate Open Graph tags

        Args:
            title: Page title
            description: Page description
            image: Featured image URL
            url: Page URL

        Returns:
            Dictionary of Open Graph tags
        """
        og_tags = {
            "og:title": title,
            "og:description": description,
            "og:type": "website",
            "og:site_name": self.site_name,
        }

        if image:
            og_tags["og:image"] = image
            og_tags["og:image:alt"] = title

        if url:
            og_tags["og:url"] = url

        return og_tags

    def _generate_twitter_card_tags(
        self,
        title: str,
        description: str,
        image: Optional[str] = None,
    ) -> Dict[str, str]:
        """Generate Twitter Card tags

        Args:
            title: Page title
            description: Page description
            image: Featured image URL

        Returns:
            Dictionary of Twitter Card tags
        """
        twitter_tags = {
            "twitter:card": "summary_large_image",
            "twitter:title": title,
            "twitter:description": description,
        }

        if image:
            twitter_tags["twitter:image"] = image

        return twitter_tags

    def generate_schema_org_data(
        self,
        schema_type: str,
        data: Dict[str, Any],
    ) -> Dict[str, Any]:
        """Generate Schema.org structured data

        Args:
            schema_type: Type of schema (Article, BlogPosting, WebSite, etc.)
            data: Data for the schema

        Returns:
            Schema.org structured data as dictionary
        """
        schema = {
            "@context": "https://schema.org",
            "@type": schema_type,
        }

        schema.update(data)

        # Add common fields
        if schema_type in ["Article", "BlogPosting"]:
            schema.setdefault("publisher", {
                "@type": "Organization",
                "name": self.site_name,
                "url": self.site_url,
            })

        return schema

    def generate_article_schema(
        self,
        title: str,
        description: str,
        url: str,
        image: Optional[str] = None,
        author_name: Optional[str] = None,
        published_date: Optional[datetime] = None,
        modified_date: Optional[datetime] = None,
    ) -> Dict[str, Any]:
        """Generate schema for an article/blog post

        Args:
            title: Article title
            description: Article description
            url: Article URL
            image: Featured image URL
            author_name: Author name
            published_date: Publication date
            modified_date: Last modified date

        Returns:
            Schema.org structured data
        """
        data = {
            "headline": title,
            "description": description,
            "url": url,
        }

        if image:
            data["image"] = image

        if author_name:
            data["author"] = {
                "@type": "Person",
                "name": author_name,
            }

        if published_date:
            data["datePublished"] = published_date.isoformat()

        if modified_date:
            data["dateModified"] = modified_date.isoformat()

        return self.generate_schema_org_data("BlogPosting", data)

    def generate_breadcrumb_schema(
        self,
        breadcrumbs: List[Dict[str, str]],
    ) -> Dict[str, Any]:
        """Generate schema for breadcrumbs

        Args:
            breadcrumbs: List of breadcrumb items with 'name' and 'url'

        Returns:
            Schema.org structured data for breadcrumbs
        """
        items = []
        for i, crumb in enumerate(breadcrumbs, 1):
            items.append({
                "@type": "ListItem",
                "position": i,
                "name": crumb["name"],
                "item": crumb["url"],
            })

        return self.generate_schema_org_data("BreadcrumbList", {
            "itemListElement": items,
        })

    def generate_robots_txt(
        self,
        disallow_paths: Optional[List[str]] = None,
        allow_paths: Optional[List[str]] = None,
        sitemap_url: Optional[str] = None,
    ) -> str:
        """Generate robots.txt content

        Args:
            disallow_paths: Paths to disallow
            allow_paths: Paths to allow (overrides disallow)
            sitemap_url: Sitemap URL

        Returns:
            robots.txt content
        """
        lines = ["User-agent: *"]

        # Allow specific paths
        if allow_paths:
            for path in allow_paths:
                lines.append(f"Allow: {path}")

        # Disallow specific paths
        if disallow_paths:
            for path in disallow_paths:
                lines.append(f"Disallow: {path}")

        # Add sitemap
        if sitemap_url:
            lines.append(f"\nSitemap: {sitemap_url}")

        return "\n".join(lines)

    def generate_sitemap_entry(
        self,
        url: str,
        last_modified: Optional[datetime] = None,
        change_frequency: Optional[str] = None,
        priority: Optional[float] = None,
    ) -> Dict[str, Any]:
        """Generate a single sitemap entry

        Args:
            url: Page URL
            last_modified: Last modified date
            change_frequency: Change frequency (always, hourly, daily, weekly, monthly, yearly, never)
            priority: Priority (0.0 to 1.0)

        Returns:
            Sitemap entry dictionary
        """
        entry = {"loc": url}

        if last_modified:
            entry["lastmod"] = last_modified.isoformat()

        if change_frequency:
            entry["changefreq"] = change_frequency

        if priority is not None:
            entry["priority"] = str(priority)

        return entry

    def generate_sitemap_xml(
        self,
        entries: List[Dict[str, Any]],
    ) -> str:
        """Generate sitemap XML

        Args:
            entries: List of sitemap entries

        Returns:
            Sitemap XML string
        """
        xml_lines = [
            '<?xml version="1.0" encoding="UTF-8"?>',
            '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
        ]

        for entry in entries:
            xml_lines.append("  <url>")
            xml_lines.append(f"    <loc>{entry['loc']}</loc>")

            if "lastmod" in entry:
                xml_lines.append(f"    <lastmod>{entry['lastmod']}</lastmod>")

            if "changefreq" in entry:
                xml_lines.append(f"    <changefreq>{entry['changefreq']}</changefreq>")

            if "priority" in entry:
                xml_lines.append(f"    <priority>{entry['priority']}</priority>")

            xml_lines.append("  </url>")

        xml_lines.append("</urlset>")

        return "\n".join(xml_lines)

    def slugify(self, text: str) -> str:
        """Convert text to URL-friendly slug

        Args:
            text: Text to slugify

        Returns:
            URL-friendly slug
        """
        import re
        import unicodedata

        # Normalize unicode characters
        text = unicodedata.normalize('NFKD', text).encode('ascii', 'ignore').decode('utf-8')

        # Convert to lowercase and replace spaces with hyphens
        text = text.lower()
        text = re.sub(r'[^\w\s-]', '', text)
        text = re.sub(r'[-\s]+', '-', text)

        return text.strip('-')

    def extract_keywords(self, content: str, max_keywords: int = 10) -> List[str]:
        """Extract keywords from content

        Args:
            content: Content to extract keywords from
            max_keywords: Maximum number of keywords to extract

        Returns:
            List of keywords
        """
        from collections import Counter
        import re

        # Common words to ignore
        stop_words = {
            'the', 'a', 'an', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for',
            'of', 'with', 'by', 'from', 'as', 'is', 'was', 'are', 'were', 'been',
            'be', 'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would',
            'could', 'should', 'may', 'might', 'must', 'can', 'this', 'that',
            'these', 'those', 'i', 'you', 'he', 'she', 'it', 'we', 'they',
        }

        # Extract words
        words = re.findall(r'\b[a-z]{3,}\b', content.lower())

        # Filter stop words
        words = [w for w in words if w not in stop_words]

        # Count occurrences
        word_counts = Counter(words)

        # Get top keywords
        keywords = [word for word, count in word_counts.most_common(max_keywords)]

        return keywords


# Global SEO service instance
seo_service = SEOService()

File: frontend/services/seo-service.ts
/**
 * SEO 优化服务
 * 提供 SEO 相关的工具函数和组件
 */

export interface SEOConfig {
  title: string;
  description: string;
  keywords?: string[];
  ogImage?: string;
  ogType?: 'website' | 'article';
  twitterCard?: 'summary' | 'summary_large_image';
  canonical?: string;
  noindex?: boolean;
  nofollow?: boolean;
  structuredData?: object;
}

export interface MetaTag {
  name?: string;
  property?: string;
  content: string;
}

export interface JsonLdProps {
  type: string;
  data: object;
}

class SEOService {
  private defaultConfig: Partial<SEOConfig> = {
    ogType: 'website',
    twitterCard: 'summary_large_image',
  };

  /**
   * 生成 meta 标签
   */
  generateMetaTags(config: SEOConfig): MetaTag[] {
    const tags: MetaTag[] = [];
    const fullConfig = { ...this.defaultConfig, ...config };

    // 基础 meta 标签
    tags.push({ name: 'title', content: fullConfig.title });
    tags.push({ name: 'description', content: fullConfig.description });

    // Keywords
    if (fullConfig.keywords && fullConfig.keywords.length > 0) {
      tags.push({ name: 'keywords', content: fullConfig.keywords.join(', ') });
    }

    // Robots
    if (fullConfig.noindex || fullConfig.nofollow) {
      const robots = [fullConfig.noindex ? 'noindex' : 'index', fullConfig.nofollow ? 'nofollow' : 'follow']
        .filter(Boolean)
        .join(', ');
      tags.push({ name: 'robots', content: robots });
    }

    // Canonical
    if (fullConfig.canonical) {
      tags.push({ name: 'canonical', content: fullConfig.canonical });
    }

    // Open Graph
    tags.push({ property: 'og:title', content: fullConfig.title });
    tags.push({ property: 'og:description', content: fullConfig.description });
    tags.push({ property: 'og:type', content: fullConfig.ogType! });

    if (fullConfig.ogImage) {
      tags.push({ property: 'og:image', content: fullConfig.ogImage });
    }

    // Twitter Card
    tags.push({ name: 'twitter:card', content: fullConfig.twitterCard! });
    tags.push({ name: 'twitter:title', content: fullConfig.title });
    tags.push({ name: 'twitter:description', content: fullConfig.description });

    if (fullConfig.ogImage) {
      tags.push({ name: 'twitter:image', content: fullConfig.ogImage });
    }

    return tags;
  }

  /**
   * 生成 JSON-LD 结构化数据
   */
  generateJsonLd(type: string, data: object): string {
    const jsonLd = {
      '@context': 'https://schema.org',
      '@type': type,
      ...data,
    };

    return JSON.stringify(jsonLd);
  }

  /**
   * 生成文章结构化数据
   */
  generateArticleSchema(data: {
    title: string;
    description: string;
    image: string;
    datePublished: string;
    dateModified?: string;
    author: {
      name: string;
      url?: string;
    };
    publisher: {
      name: string;
      logo?: string;
    };
    url?: string;
  }): string {
    return this.generateJsonLd('Article', {
      headline: data.title,
      description: data.description,
      image: data.image,
      datePublished: data.datePublished,
      dateModified: data.dateModified || data.datePublished,
      author: {
        '@type': 'Person',
        name: data.author.name,
        url: data.author.url,
      },
      publisher: {
        '@type': 'Organization',
        name: data.publisher.name,
        logo: data.publisher.logo,
      },
      mainEntityOfPage: data.url,
    });
  }

  /**
   * 生成面包屑结构化数据
   */
  generateBreadcrumbSchema(items: Array<{ name: string; url: string }>): string {
    return this.generateJsonLd('BreadcrumbList', {
      itemListElement: items.map((item, index) => ({
        '@type': 'ListItem',
        position: index + 1,
        name: item.name,
        item: item.url,
      })),
    });
  }

  /**
   * 生成网站结构化数据
   */
  generateWebsiteSchema(data: {
    name: string;
    url: string;
    description?: string;
    searchAction?: string;
  }): string {
    const schema: any = {
      '@type': 'WebSite',
      name: data.name,
      url: data.url,
    };

    if (data.description) {
      schema.description = data.description;
    }

    if (data.searchAction) {
      schema.potentialAction = {
        '@type': 'SearchAction',
        target: `${data.searchAction}?search={search_term_string}`,
        'query-input': 'required name=search_term_string',
      };
    }

    return this.generateJsonLd('WebSite', schema);
  }

  /**
   * 生成组织结构化数据
   */
  generateOrganizationSchema(data: {
    name: string;
    url: string;
    logo?: string;
    description?: string;
    sameAs?: string[];
    contactPoint?: {
      type: string;
      telephone?: string;
      email?: string;
    };
  }): string {
    const schema: any = {
      '@type': 'Organization',
      name: data.name,
      url: data.url,
    };

    if (data.logo) {
      schema.logo = data.logo;
    }

    if (data.description) {
      schema.description = data.description;
    }

    if (data.sameAs && data.sameAs.length > 0) {
      schema.sameAs = data.sameAs;
    }

    if (data.contactPoint) {
      schema.contactPoint = {
        '@type': data.contactPoint.type,
        contactType: 'customer service',
      };

      if (data.contactPoint.telephone) {
        schema.contactPoint.telephone = data.contactPoint.telephone;
      }

      if (data.contactPoint.email) {
        schema.contactPoint.email = data.contactPoint.email;
      }
    }

    return this.generateJsonLd('Organization', schema);
  }

  /**
   * 生成产品结构化数据
   */
  generateProductSchema(data: {
    name: string;
    description: string;
    image: string[];
    brand?: string;
    sku?: string;
    offers: {
      price: number;
      priceCurrency: string;
      availability: 'InStock' | 'OutOfStock' | 'PreOrder';
      url?: string;
    };
  }): string {
    return this.generateJsonLd('Product', {
      name: data.name,
      description: data.description,
      image: data.image,
      brand: data.brand
        ? {
            '@type': 'Brand',
            name: data.brand,
          }
        : undefined,
      sku: data.sku,
      offers: {
        '@type': 'Offer',
        price: data.offers.price,
        priceCurrency: data.offers.priceCurrency,
        availability: `https://schema.org/${data.offers.availability}`,
        url: data.offers.url,
      },
    });
  }

  /**
   * 优化图片 URL（用于 SEO）
   */
  optimizeImageUrl(url: string, options?: { width?: number; height?: number; quality?: number }): string {
    if (!url) return '';

    try {
      const urlObj = new URL(url, typeof window !== 'undefined' ? window.location.origin : '');

      // 添加图片优化参数
      if (options?.width) {
        urlObj.searchParams.set('w', options.width.toString());
      }
      if (options?.height) {
        urlObj.searchParams.set('h', options.height.toString());
      }
      if (options?.quality) {
        urlObj.searchParams.set('q', options.quality.toString());
      }

      return urlObj.toString();
    } catch (e) {
      return url;
    }
  }

  /**
   * 生成 slug（URL 友好的字符串）
   */
  generateSlug(text: string): string {
    return text
      .toLowerCase()
      .trim()
      .replace(/[^\w\s-]/g, '')
      .replace(/[\s_-]+/g, '-')
      .replace(/^-+|-+$/g, '');
  }

  /**
   * 生成页面标题
   */
  generatePageTitle(title: string, siteName?: string, template?: string): string {
    if (template) {
      return template.replace('%s', title).replace('%siteName%', siteName || '');
    }

    if (siteName) {
      return `${title} | ${siteName}`;
    }

    return title;
  }

  /**
   * 生成元描述
   */
  generateMetaDescription(content: string, maxLength: number = 160): string {
    // 移除 HTML 标签
    const text = content.replace(/<[^>]*>/g, '');

    // 截断到指定长度
    if (text.length <= maxLength) {
      return text;
    }

    // 在单词边界截断
    const truncated = text.substr(0, maxLength);
    const lastSpace = truncated.lastIndexOf(' ');

    return truncated.substr(0, lastSpace) + '...';
  }

  /**
   * 生成关键词
   */
  generateKeywords(content: string, stopWords?: string[]): string[] {
    const text = content.toLowerCase();
    const words = text.match(/\b\w+\b/g) || [];

    const defaultStopWords = [
      'a',
      'an',
      'the',
      'and',
      'or',
      'but',
      'in',
      'on',
      'at',
      'to',
      'for',
      'of',
      'with',
      'by',
      'is',
      'are',
      'was',
      'were',
      'be',
      'been',
      'being',
    ];

    const stopWordsSet = new Set([...defaultStopWords, ...(stopWords || [])]);

    // 统计词频
    const wordCount = new Map<string, number>();
    words.forEach((word) => {
      if (word.length > 2 && !stopWordsSet.has(word)) {
        wordCount.set(word, (wordCount.get(word) || 0) + 1);
      }
    });

    // 返回最常见的词
    return Array.from(wordCount.entries())
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10)
      .map(([word]) => word);
  }

  /**
   * 验证 SEO 配置
   */
  validateSEOConfig(config: SEOConfig): { valid: boolean; errors: string[] } {
    const errors: string[] = [];

    if (!config.title || config.title.length === 0) {
      errors.push('Title is required');
    }

    if (config.title && config.title.length > 60) {
      errors.push('Title should be 60 characters or less');
    }

    if (!config.description || config.description.length === 0) {
      errors.push('Description is required');
    }

    if (config.description && config.description.length > 160) {
      errors.push('Description should be 160 characters or less');
    }

    if (config.ogImage && !this.isValidUrl(config.ogImage)) {
      errors.push('OG image must be a valid URL');
    }

    return {
      valid: errors.length === 0,
      errors,
    };
  }

  /**
   * 验证 URL
   */
  private isValidUrl(url: string): boolean {
    try {
      new URL(url);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * 生成 sitemap URL
   */
  generateSitemapUrl(
    url: string,
    options?: {
      lastModified?: Date;
      changeFrequency?: 'always' | 'hourly' | 'daily' | 'weekly' | 'monthly' | 'yearly' | 'never';
      priority?: number;
    }
  ): string {
    const attrs = [
      `loc="${this.escapeXml(url)}"`,
      options?.lastModified ? `lastmod="${options.lastModified.toISOString()}"` : '',
      options?.changeFrequency ? `changefreq="${options.changeFrequency}"` : '',
      options?.priority !== undefined ? `priority="${options.priority}"` : '',
    ].filter(Boolean);

    return `<url>${attrs.map((attr) => `<${attr} />`).join('')}</url>`;
  }

  /**
   * 转义 XML
   */
  private escapeXml(str: string): string {
    return str
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&apos;');
  }
}

// 创建全局实例
export const seoService = new SEOService();

// Hook
export function useSEO() {
  const generateMetaTags = React.useCallback((config: SEOConfig) => {
    return seoService.generateMetaTags(config);
  }, []);

  const generateArticleSchema = React.useCallback(
    (data: Parameters<typeof seoService.generateArticleSchema>[0]) => {
      return seoService.generateArticleSchema(data);
    },
    []
  );

  const generateSlug = React.useCallback((text: string) => {
    return seoService.generateSlug(text);
  }, []);

  const validateSEOConfig = React.useCallback((config: SEOConfig) => {
    return seoService.validateSEOConfig(config);
  }, []);

  return {
    generateMetaTags,
    generateArticleSchema,
    generateSlug,
    validateSEOConfig,
  };
}

export default SEOService;
